apps/voting-disputable/contracts/DisputableVoting.sol

```diff
@@ -0,0 +1,745 @@
+
+    /**
+    * @dev Internal function to create a new vote
+    * @return voteId id for newly created vote
+    */
+    function _newVote(bytes _executionScript, string _metadata, bool _castVote) internal returns (uint256 voteId) {
+        uint64 snapshotBlock = getBlockNumber64() - 1; // avoid double voting in this very block
+        uint256 votingPower = token.totalSupplyAt(snapshotBlock);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
+        voteId = votesLength++;
+
+        Vote storage vote_ = votes[voteId];
+        vote_.startDate = getTimestamp64();
+        vote_.snapshotBlock = snapshotBlock;
+        vote_.overruleWindow = overruleWindow;
+        vote_.supportRequiredPct = supportRequiredPct;
+        vote_.minAcceptQuorumPct = minAcceptQuorumPct;
+        vote_.votingPower = votingPower;
+        vote_.executionScript = _executionScript;
+        vote_.actionId = _newAgreementAction(voteId, msg.sender, bytes(_metadata));
+        vote_.disputableStatus = DisputableStatus.Active;
+
+        emit StartVote(voteId, msg.sender, _metadata);
+
+        if (_castVote && _canVote(vote_, msg.sender)) {
+            _vote(voteId, true, msg.sender);
+        }
+    }
+
+    /**
+    * @dev Internal function to cast a vote. It assumes the queried vote exists.
+    */
+    function _vote(uint256 _voteId, bool _supports, address _voter) internal {
+        // This could re-enter, though we can assume the governance token is not malicious
+        Vote storage vote_ = votes[_voteId];
+        _castVote(vote_, _voteId, _supports, _voter);
+        _overwriteCasterIfNecessary(vote_, _voter);
+
+        if (_canExecute(vote_)) {
+            // We've already checked if the vote can be executed with `_canExecute()`
+            _unsafeExecuteVote(_voteId);
+        }
+    }
+
+    /**
+    * @dev Internal function for a representative to cast a vote on behalf of many voters. It assumes the queried vote exists.
+    */
+    function _voteOnBehalfOf(address[] _voters, uint256 _voteId, bool _supports) internal {
+        Vote storage vote_ = votes[_voteId];
+        for (uint256 i = 0; i < _voters.length; i++) {
+            address voter = _voters[i];
+            require(_canVote(vote_, voter), ERROR_CAN_NOT_VOTE);
+            require(!_withinOverruleWindow(vote_), ERROR_WITHIN_OVERRULE_WINDOW);
+            require(_isRepresentativeOf(voter, msg.sender), ERROR_NOT_REPRESENTATIVE);
+
+            if (_hasNotVotedYet(vote_, voter)) {
+                _voteOnBehalfOf(_voteId, _supports, voter, msg.sender);
+            } else {
+                emit ProxyVoteFailure(_voteId, voter, msg.sender);
+            }
+        }
+    }
+
+    /**
+    * @dev Internal function for a representative to cast a vote on behalf of a voter. It assumes the queried vote exists.
+    */
+    function _voteOnBehalfOf(uint256 _voteId, bool _supports, address _voter, address _representative) internal {
+        Vote storage vote_ = votes[_voteId];
```

Comment:
It's kind of weird this has a very different signature than the other one.

Another thing I'm thinking about is just taking the `Vote` as a storage variable, as extra SLOADs are so expensive now.

Reply:
Totally!